core/unit_test/view/TestViewCustomizationAccessorFromMapping.hpp

```diff
@@ -0,0 +1,215 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#include <gtest/gtest.h>
+
+namespace Foo {
+
+// ElementType which is actually just a tag type
+struct BarStrided {};
+
+// Reference type for Bar
+struct BarRefStrided {
+  double* ptr{nullptr};
+  size_t size{0ul};
+  size_t stride{0ul};
+  KOKKOS_FUNCTION
+  double& operator[](size_t idx) const { return ptr[idx * stride]; }
+};
+
+// A TestAccessor mimicking some of what Sacado does
+// Specifically:
+//   * returns a proxy reference
+//   * the underlying storage is of some basic scalar type
+//   * the size of the allocation does not actually come from
+//   sizeof(ElementType)
+template <class ElementType, class MemorySpace>
+struct TestAccessorStrided {
+  static_assert(std::is_same_v<std::remove_cv_t<ElementType>, BarStrided>);
+  using element_type     = ElementType;
+  using reference        = BarRefStrided;
+  using data_handle_type = Kokkos::Impl::ReferenceCountedDataHandle<
+      std::conditional_t<std::is_const_v<ElementType>, const double, double>,
+      MemorySpace>;
+  using offset_policy = TestAccessorStrided;
+
+  // View expects this from accessors right now
+  using memory_space = MemorySpace;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  constexpr TestAccessorStrided() = default;
+
+  template <class OtherElementType,
+            std::enable_if_t<std::is_constructible_v<
+                                 Kokkos::default_accessor<element_type>,
+                                 Kokkos::default_accessor<OtherElementType>>,
+                             int> = 0>
+  KOKKOS_FUNCTION constexpr TestAccessorStrided(
+      const TestAccessorStrided<OtherElementType, MemorySpace>& other) noexcept
+      : size(other.size), stride(other.stride) {}
+
+  KOKKOS_FUNCTION
+  TestAccessorStrided(const size_t val, const size_t s_val)
+      : size(val), stride(s_val) {}
+
+  KOKKOS_FUNCTION
+  constexpr reference access(
+#ifndef KOKKOS_ENABLE_OPENACC
+      const data_handle_type& p,
+#else
+      // FIXME OPENACC: illegal address when passing by reference
+      data_handle_type p,
+#endif
+      size_t i) const noexcept {
+    return BarRefStrided{(p.get() + i), size, stride};
+  }
+
+  KOKKOS_FUNCTION
+  constexpr typename offset_policy::data_handle_type offset(
+#ifndef KOKKOS_ENABLE_OPENACC
+      const data_handle_type& p,
+#else
+      // FIXME OPENACC: illegal address when passing by reference
+      data_handle_type p,
+#endif
```

Comment:
It's a pointer; would you consider just passing it by value and avoiding the macrology?  Then you wouldn't need a special case.

```suggestion
      data_handle_type p,
```

Reply:
oh it actually isn't. In our case it's more often than not a custom implementation of shared ptr. and we definitely want to avoid ref counting here.

Reply:
oh, it's a `shared_ptr` something, never mind : - )